### PR TITLE
Add option for keeping local cache of recent data. 

### PR DESCRIPTION
**Pull request by trueleo**
### Description

This PR adds `local cache` / `hot tier` for Parseable.

This option can be enabled by setting following env vars
`P_CACHE_DIR` - Local Path for file cache
`P_CACHE_SIZE` - Size for cache in human readable size ( mb/mib/gib/gb )

When these flags are set, sync flow will move the parquet from staging into the specified cache directory instead of deleting it. Any LRU cached entry is deleted to satisfy the cache constraint upon insertion. LocalCacheManager is responsible for updating and persisting this data structure.

<hr>

This PR has:
- [ ] been tested to ensure log ingestion and log query works.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added documentation for new or modified features or behaviors.

### CODE REVIEW

File: server/src/option.rs [hunk 1]
```diff
@@ -384,6 +398,25 @@ impl Server {
...
+                    .next_line_help(true),
+            )
+             .arg(
+                Arg::new(Self::CACHE_SIZE)
+                    .long(Self::CACHE_SIZE)
+                    .env("P_CACHE_SIZE")
+                    .value_name("mb")
+                    .default_value("1000")
+                    .value_parser(value_parser!(usize))
+                    .help("The local staging path is used as a temporary landing point for incoming events and local cache")
```
Comment (nitisht) [inline server/src/option.rs:418]: Help text is wrong

File: server/src/option.rs [hunk 2]
```diff
@@ -384,6 +398,25 @@ impl Server {
...
+                    .value_parser(validation::canonicalize_path)
+                    .help("The local staging path is used as a temporary landing point for incoming events and local cache")
+                    .next_line_help(true),
+            )
+             .arg(
+                Arg::new(Self::CACHE_SIZE)
+                    .long(Self::CACHE_SIZE)
+                    .env("P_CACHE_SIZE")
+                    .value_name("mb")
+                    .default_value("1000")
```
Comment (nitisht) [inline server/src/option.rs:416]: 1000MB?

File: server/src/query/table_provider.rs [hunk 3]
```diff
@@ -1,140 +0,0 @@
-/*
```
Comment (nitisht) [inline server/src/query/table_provider.rs:1]: Why deleted?

**Review by nitisht — CHANGES_REQUESTED** [review]
Fix the commit message with more details on why we added this and how it works.

**Comment by nitisht** [comment]
Missing items

- Prometheus stats on cache as we discussed.
- Cache info printed on the startup banner under storage.

**Comment by nitisht** [comment]
What does this mapping in the .cache.json file mean?

```
[ 
"home/nitish/Development/parseablehq/parseable/data/teststream/date=2023-12-24/hour=04/minute=57/nitish-x1.data.parquet",
"/home/nitish/Development/parseablehq/parseable/cache/teststream/date=2023-12-24.hour=04.minute=57.nitish-x1.data.parquet"
]
```

**Comment by nitisht** [comment]
If I restart the server and change the cache size via `P_CACHE_SIZE`, the `.cache.json` still has old value. We need to update the `.cache.json` so current specified size is honoured.

**Comment by theteachr** [comment]
> What does this mapping in the .cache.json file mean?

@nitisht It's a mapping from store path to where that is stored in the local cache.

**Comment by nitisht** [comment]
> What does this mapping in the .cache.json file mean?
> 
> ```
> [ 
> "home/nitish/Development/parseablehq/parseable/data/teststream/date=2023-12-24/hour=04/minute=57/nitish-x1.data.parquet",
> "/home/nitish/Development/parseablehq/parseable/cache/teststream/date=2023-12-24.hour=04.minute=57.nitish-x1.data.parquet"
> ]
> ```



> > What does this mapping in the .cache.json file mean?
> 
> @nitisht It's a mapping from store path to where the file is stored in the local cache.

That is clear, but

- Both files are local, still there is difference in path format - leading / vs no /
- Where do we use this mapping - that info should be added as comments and commit message

File: server/src/query/table_provider.rs (hunk 3 above)
Comment (trueleo) [reply server/src/query/table_provider.rs:1]: This file is unused now

Comment (nitisht) [reply server/src/query/table_provider.rs:1]: Let's create a separate cleanup PR with details on why files are not used. this change is clearly unrelated to caching.

Comment (trueleo) [reply server/src/query/table_provider.rs:1]: >  This change is clearly unrelated to caching.

This file was being used to combine two execution plans ( remote and memory ), but now because of conditional addition of caching, I had to refactor related code in the schema provider file.

**Comment by trueleo** [comment]
> Missing items
> 
>     * Prometheus stats on cache as we discussed.
> 
>     * Cache info printed on the startup banner under storage.

I will do this in a separate PR.

File: server/src/query/table_provider.rs (hunk 3 above)
Comment (nitisht) [reply server/src/query/table_provider.rs:1]: Please fix commit message with details

**Comment by nitisht** [comment]
> > Missing items
> > ```
> > * Prometheus stats on cache as we discussed.
> > 
> > * Cache info printed on the startup banner under storage.
> > ```
> 
> I will do this in a separate PR.

Let's finish in this PR

File: server/src/option.rs [hunk 4]
```diff
@@ -606,4 +641,19 @@ pub mod validation {
...
+            s: &str,
+        ) -> Result<u64, human_size::ParsingError> {
+            SpecificSize::<T>::from_str(s).map(|x| x.to_bytes())
+        }
+
+        parse_and_map::<multiples::Mebibyte>(s)
+            .or(parse_and_map::<multiples::Megabyte>(s))
+            .or(parse_and_map::<multiples::Gigibyte>(s))
+            .or(parse_and_map::<multiples::Gigabyte>(s))
+            .map_err(|_| "Could not parse given size".to_string())
```
Comment (nitisht) [inline server/src/option.rs:657]: let's add TB and Bytes as well.

File: server/src/storage/s3.rs [hunk 5]
```diff
@@ -200,6 +200,12 @@ impl ObjectStorageProvider for S3Config {
         })
     }
 
+    fn get_store(&self) -> Arc<dyn ObjectStore> {
+        let s3 = self.get_default_builder().build().unwrap();
+        // limit objectstore to a concurrent request limit
+        Arc::new(LimitStore::new(s3, super::MAX_OBJECT_STORE_REQUESTS))
+    }
```
Comment (nitisht) [inline server/src/storage/s3.rs:207]: Why is this needed - don't it being used?

File: server/src/localcache.rs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    files: Cache<String, PathBuf>,
+}
+
+impl LocalCache {
+    fn new_with_size(capacity: u64) -> Self {
+        Self {
+            version: "v1".to_string(),
+            current_size: 0,
+            capacity,
+            files: Cache::new(100),
```
Comment (nitisht) [inline server/src/localcache.rs:47]: Why 100 only here?

**Comment by nitisht** [comment]
- Prometheus stats on cache as we discussed.
- Cache info printed on the startup banner under storage.
- Both files are local, still there is difference in path format - leading `/` vs no `/`
- Where do we use this mapping - that info should be added as comments and commit message
- If I restart the server and change the cache size via P_CACHE_SIZE, the .cache.json still has old value. We need to update the .cache.json so current specified size is honoured.
- What is the minimum allowed value for `P_CACHE_SIZE`? We need to fail if smaller values are set. IMO `1GiB` is a good number.

File: server/src/localcache.rs (hunk 6 above)
Comment (trueleo) [reply server/src/localcache.rs:47]: backing hashlru is capacity based. The capacity is arbitrary and is increased when needed.

Comment (nitisht) [reply server/src/localcache.rs:47]: who increases? user or the backing cache?

**Review by nitisht — APPROVED** [review]
